core/src/test/java/io/parsingdata/metal/data/callback/CallbackTest.java

```diff
@@ -154,9 +159,9 @@ protected void handleFailure(Token token, Environment before) {}
 
     @Test
     public void genericCallback() throws IOException {
-        final Deque<Token> expectedSuccessDefinitions = new ArrayDeque<>(Arrays.asList(ONE, TWO, ONE, TWO, FOUR, SEQ124, CHOICE));
-        final Deque<Long> expectedSuccessOffsets = new ArrayDeque<>(Arrays.asList(1L, 2L, 1L, 2L, 3L, 3L, 3L));
-        final Deque<Token> expectedFailureDefinitions = new ArrayDeque<>(Arrays.asList(THREE, SEQ123));
+        final Deque<Token> expectedSuccessDefinitions = new ArrayDeque<>(Arrays.asList(DEF_ONE, POST_ONE, DEF_TWO, POST_TWO, DEF_THREE, DEF_ONE, POST_ONE, DEF_TWO, POST_TWO, DEF_FOUR, POST_FOUR, SEQ124, CHOICE));
+        final Deque<Long> expectedSuccessOffsets = new ArrayDeque<>(Arrays.asList(1L, 1L, 2L, 2L, 3L, 1L, 1L, 2L, 2L, 3L, 3L, 3L, 3L));
+        final Deque<Token> expectedFailureDefinitions = new ArrayDeque<>(Arrays.asList(POST_THREE, SEQ123));
```

Comment:
These lists of expected success and failure definitions are nice, it illustrates what happens during parsing.

We *could* make it extremely explicit and create a list/chain of expectations that combines both success/failure and the offset where those occur, to show when during the parse run.

Something like:

```java
expectSuccess(DEF_ONE, 1), // token, offset
expectSuccess(POST_ONE, 1), 
expectSuccess(DEF_TWO, 2), 
expectSuccess(POST_TWO, 2), 
expectSuccess(DEF_THREE, 3), 
expectFailure(POST_THREE, 3), 
expectFailure(SEQ123, 3),
expectSuccess(DEF_ONE, 1), 
expectSuccess(POST_ONE, 1), 
expectSuccess(DEF_TWO, 2), 
expectSuccess(POST_TWO, 2), 
expectSuccess(DEF_FOUR, 3), 
expectSuccess(POST_FOUR, 3), 
expectSuccess(SEQ124, 3), 
expectSuccess(CHOICE, 3);
```

Food for thought.